Content.Shared/Teleportation/Systems/SharedPortalSystem.cs

```diff
@@ -145,7 +151,102 @@ private void OnCollide(EntityUid uid, PortalComponent component, ref StartCollid
 
         // no linked entity--teleport randomly
         if (component.RandomTeleport)
-            TeleportRandomly(uid, subject, component);
+            TeleportRandomly(portal, subject, component);
+    }
+
+    private void OnCanDragDropOn(EntityUid uid, PortalComponent component, ref CanDropTargetEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        args.CanDrop = CanInsert(uid, component, args.Dragged);
+        args.Handled = true;
+    }
+
+    private bool TryInsert(EntityUid uid, EntityUid toInsertId, EntityUid? userId, PortalComponent? component = null)
+    {
+        if (!Resolve(uid, ref component))
+            return false;
+
+        if (!CanInsert(uid, component, toInsertId))
+            return false;
+
+        bool insertingSelf = userId == toInsertId;
+
+        var delay = insertingSelf ? component.EntryDelay : component.DraggedEntryDelay;
+
+        if (userId != null && !insertingSelf)
+            _popup.PopupEntity(Loc.GetString("portal-being-inserted", ("user", Identity.Entity((EntityUid) userId, EntityManager))), toInsertId, toInsertId, PopupType.Large);
+
+        if (delay <= 0 || userId == null)
```

Comment:
This should use popuppredicted or it's going to spam popups with prediction.

Reply:
I went with `PopupClient` instead.